pkg/controllers/provisioning/v1alpha1/reallocator/filter.go

```diff
@@ -25,40 +30,81 @@ type Filter struct {
 	kubeClient client.Client
 }
 
-// Gets Nodes fitting some underutilization
-// TODO: add predicate
-func (f *Filter) GetUnderutilizedNodes(ctx context.Context) ([]*v1.Node, error) {
-	nodeList := &v1.NodeList{}
+// Gets Nodes fitting some underutilization predicate
+func (f *Filter) GetUnderutilizedNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) ([]*v1.Node, error) {
 	underutilized := []*v1.Node{}
 
-	// 1. Get all nodes
-	if err := f.kubeClient.List(ctx, nodeList); err != nil {
-		return nil, fmt.Errorf("listing nodes, %w", err)
+	// 1. Get all provisioner labeled nodes
+	nodes, err := f.getNodes(ctx, provisioner)
+	if err != nil {
+		return nil, err
 	}
 
 	// 2. Get nodes and the pods on each node
-	for _, node := range nodeList.Items {
+	for _, node := range nodes {
 		pods, err := f.getPodsOnNode(ctx, node.Name)
 		if err != nil {
-			return []*v1.Node{}, fmt.Errorf("filtering pods on nodes, %w", err)
+			zap.S().Debugf("unable to get pods for node %s, %s", node.Name, err.Error())
+			continue
 		}
-		if f.isUnderutilized(&node, pods) {
-			underutilized = append(underutilized, &node)
+
+		// Only checks if it has 0 non-daemon pods right now
+		if f.isUnderutilized(pods) {
+			underutilized = append(underutilized, node)
 		}
 	}
 	return underutilized, nil
 }
 
+func (f *Filter) GetExpiredNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) ([]*v1.Node, error) {
+	nodes := []*v1.Node{}
```

Comment:
I'd define this list just before it's used on line 66. I'd also call this expiredNodes so that you don't have to use the awkward form nodeList. Currently it's not clear what the difference is between nodes and nodeList to a reader without reading the code.